File: src/network.py
from collections import deque
from link import Link
from node import Node
from relationship import Relationship


class Network(object):

    def __init__(self):
        self._nodes = []
        self._ids = {}
        self._netid_count = -1

    @staticmethod
    def from_file(path):
        keymap = {'3': Relationship.P, '2': Relationship.R, '1': Relationship.C}
        net = Network()
        with open(path) as file:
            for line in file:
                words = line.split()
                net.add_link(int(words[0]), int(words[1]), keymap[words[2]])
        return net

    def _new_netid(self):
        self._netid_count += 1
        return self._netid_count

    def _netid(self, id):
        try:
            netid = self._ids[id]
        except KeyError:
            netid = self._new_netid()
            self._ids[id] = netid

        return netid

    def add_link(self, tail_id, head_id, relationship):
        tail_netid = self._netid(tail_id)
        head_netid = self._netid(head_id)

        # check if the tail id is not in the nodes list
        # this is true if the ID is bigger then the current list size
        if tail_netid >= len(self._nodes):
            self._nodes.append(Node(tail_id, tail_netid))

        # check if the tail id is not in the nodes list
        if head_netid >= len(self._nodes):
            self._nodes.append(Node(head_id, head_netid))

        # add the edge to the tail node
        self._nodes[tail_netid].add_link(self._nodes[head_netid], relationship)

    def find_routes(self, dest_id):

        for node in self._nodes:
            node.path_type = Relationship.NON

        # create aux list of edges
        customer_links = deque()
        peer_links = deque()
        provider_links = deque()

        only_customers = False
        dest_node = self._nodes[self._ids[dest_id]]
        link = Link(dest_node, dest_node, Relationship.SELF)

        while link:
            # processing the path type of the head node of the link
            node = link.head

            path_type = Relationship.operation(link.type, link.tail.path_type)
            if path_type < node.path_type:
                node.path_type = path_type

                for customer in node.customers:
                    # the type of link is equal to the relationship between the head node and the tail
                    customer_links.append(Link(node, customer, Relationship.C))

                if not only_customers:
                    for peer in node.peers:
                        peer_links.append(Link(node, peer, Relationship.R))
                    for provider in node.providers:
                        provider_links.append(Link(node, provider, Relationship.P))

            if provider_links:
                link = provider_links.popleft()
            elif peer_links:
                link = peer_links.popleft()
            elif customer_links:
                link = customer_links.popleft()
                only_customers = True
            else:
                link = None

    def get_hops(self, dest_id):
        min_hops = {}
        tovisit = deque()

        # Set all the distances to infinity
        for node in self._nodes:
            min_hops[node._netid] = 4294967295 # it should be infinity here

        # Map the node to our id strategy
        node = self._nodes[self._ids[dest_id]]

        # Set the distance to self equal to zero
        min_hops[node._netid] = 0

        # Add the edge to self to the queue
        link = Link(node, node, Relationship.C)

        while link:

            node = link.head

            # Check if the current estimate is better than the estimated last
            if min_hops[node._netid] > 1 + min_hops[link.tail._netid] or link.head._netid == link.tail._netid:

                # Update estimate
                if link.head._netid != link.tail._netid:
                    min_hops[node._netid] = 1 + min_hops[link.tail._netid]

                # Add all the links going from the destination node to the queue
                # Also add the links to visited links set
                for customer in node.customers:
                    if customer._netid != link.tail._netid:
                        tovisit.append(Link(node, customer, Relationship.C))

                if node.path_type == 1 or node.path_type == 0:
                    for peer in node.peers:
                        if peer._netid != link.tail._netid:
                            tovisit.append(Link(node, peer, Relationship.R))
                    for provider in node.providers:
                        if provider._netid != link.tail._netid:
                            tovisit.append(Link(node, provider, Relationship.P))

            if tovisit:
                link = tovisit.popleft()
            else:
                return min_hops

    def __repr__(self):
        representation = 'Network:\n'
        for node in self._nodes:
            representation += '(' + \
                              'id=' + str(node.id) + ', ' + \
                              'customers=' + str(node.customers) + ', ' + \
                              'peers=' + str(node.peers) + ', ' + \
                              'providers=' + str(node.providers) + ', ' + \
                              'path_type=' + str(node.path_type) + ')\n'
        return representation

File: src/Network.cpp
#include <fstream>
#include <iostream>

#include "Network.h"
#include "LinkPriorityQueue.h"

using namespace std;

Network::Network(const std::string &filename) {

    ifstream inFile(filename, ifstream::in);

    if(!inFile.is_open()) {
        throw string("can't open input file");
    }

    uint16_t tailId, headId, relationship;
    while(inFile >> tailId >> headId >> relationship) {
        addLink(tailId, headId, Link::Type (relationship - 1));
    }
}

void Network::addLink(Node::ID tailId, Node::ID headId, Link::Type  type) {

    // get the network ids for both the nodes of the link
    Node::ID headNetId = idGenerator.getNetworkId(headId);
    Node::ID tailNetId = idGenerator.getNetworkId(tailId);

    // add the nodes to the nodes list
    if(headNetId >= nodes.size()) {
        nodes.push_back(NodeHolder(new Node(headId, headNetId)));
    }

    if(tailNetId >= nodes.size()) {
        nodes.push_back(NodeHolder(new Node(tailId, tailNetId)));
    }

    switch (type) {
        case Link::Type ::Customer:
            // add  customer link from the tail node to the head node
            nodes[tailNetId]->addCustomer(nodes[headNetId].get());
            break;
        case Link::Type ::Peer:
            // add  peer link from the tail node to the head node
            nodes[tailNetId]->addPeer(nodes[headNetId].get());
            break;
        case Link::Type ::Provider:
            // add  peer provider from the tail node to the head node
            nodes[tailNetId]->addProvider(nodes[headNetId].get());
            break;
    }
}

void Network::print() const {
    for(auto& node : nodes) {
        cout << node->getId() << ":" << endl;

        for(auto& customer : node->getCustomers()) {
            cout << "\tC-" << customer->getId() << endl;
        }

        for(auto& peer : node->getPeers()) {
            cout << "\tR-" << peer->getId() << endl;
        }

        for(auto& provider : node->getProviders()) {
            cout << "\tP-" << provider->getId() << endl;
        }
    }
}

Network::PathTypeList Network::findPathTypes(Node::ID destNodeId) {

    // get the destination node network id
    Node::ID destNodeNetId = idGenerator.getNetworkId(destNodeId);

    // initialize all nodes with path type = None
    PathTypeList pathTypes(nodes.size(), PathType::None);

    LinkPriorityQueue linkQueue;
    pathTypes[destNodeNetId] = PathType::Customer;

    linkQueue.pushCustomers(nodes[destNodeNetId].get());
    linkQueue.pushPeers(nodes[destNodeNetId].get());
    linkQueue.pushProviders(nodes[destNodeNetId].get());

    while(!linkQueue.empty()) {
        Link link = linkQueue.pop();
        Node* node = link.getHead();

        PathType newPathType = operation(link.getType(), pathTypes[link.getTail()->getNetid()]);
        if(newPathType < pathTypes[node->getNetid()]) {
            pathTypes[node->getNetid()] = newPathType;

            linkQueue.pushCustomers(node);

            if(newPathType == PathType::Customer) {
                linkQueue.pushPeers(node);
                linkQueue.pushProviders(node);
            }
        }
    }

    return pathTypes;
}

Network::PathType Network::operation(Link::Type linkType, Network::PathType pathType) {
    static PathType table[3][5] = {
            {PathType::Provider, PathType::Provider, PathType::Provider, PathType::Provider, PathType::None},
            {PathType::Peer,     PathType::Peer,     PathType::None,     PathType::None,     PathType::None},
            {PathType::Customer, PathType::Customer, PathType::None,     PathType::None,     PathType::None}
    };

    return table[linkType][pathType];
}
